### PR TITLE
Integrated libx264 for higher quality mp4 output

### LINKED ISSUE #1118: Saving an animation at 4 frames per second on mac creates green animation
When I save an animation of CLT at 4 FPS on my mac, it outputs a solid green video. I've tried this both from my UI stuff as well as from scripts.

@doutriaux1

### PR DESCRIPTION
**Pull request by chaosphere2112**
Fixes #1118, and gives us *much* nicer quality animation output.

### CODE REVIEW

**Comment by durack1** [comment]
@chaosphere2112 nice! *How much* nicer are the outputs? And are they also smaller?

**Comment by doutriaux1** [comment]
@chaosphere2112 looks like you're missing some files.

**Comment by doutriaux1** [comment]
@chaosphere2112 you might as well use cdat_confiugre_step for x264 as well.

**Comment by chaosphere2112** [comment]
@doutriaux1 I just did a dumb thing (removed the ffmpeg_configure_step.cmake.in and x264_configure_step.cmake.in without removing the configure_file command, despite them not being in use). Also forgot to update the ffmpeg() call on canvas to actually take advantage of this stuff.

@durack1 A *lot* better. And, yes, smaller as well.

Codec | Size | Quality
--------- | ------ | ---------
[mpeg2](http://uvcdat.llnl.gov/mov/mp2_low.mp4) | 2 MB | Low
[mpeg2](http://uvcdat.llnl.gov/mov/mp2_high.mp4) | 6.3 MB | High
[h264](http://uvcdat.llnl.gov/mov/h264.mp4) | 3.6 MB | High

**Comment by doutriaux1** [comment]
@chaosphere2112 yasm stuff

**Comment by durack1** [comment]
@chaosphere2112 great - thanks for doing this.. I also note that the `h264/*.mp4` encoded files play directly in any modern browser, whereas the `mpeg2/*.mp4` files can be painful with codecs (and don't just play in FF41).. This is a great jump forward!

**Comment by chaosphere2112** [comment]
@doutriaux1 On [the system requirements page](https://github.com/UV-CDAT/uvcdat/wiki/System-Requirements) we list YASM as a requirement for Ubuntu, though neither of the other platforms. Hm... I can just wrap the configure args to check if YASM is installed and disable it otherwise?

File: Packages/vcs/Lib/Canvas.py [hunk 1]
```diff
@@ -4718,10 +4718,7 @@ def ffmpeg(self, movie, files, bitrate=1024, rate=None, options=''):
             cmd += '-i %s_%%d.png' % (rnd)
         elif isinstance(files, str):
             cmd += '-i ' + files
-        if rate is not None:
-            cmd += ' -r %s ' % rate
-        if bitrate is not None:
-            cmd += ' -b:v %sk' % bitrate
+        cmd += " -pix_fmt yuv420p "
```
Comment (durack1) [inline Packages/vcs/Lib/Canvas.py]: @chaosphere2112 are `h264` encoded files now the new default for the `x.ffmpeg()` and the `x.animation()` functions (where `x = vcs.init()`)?

Comment (chaosphere2112) [inline Packages/vcs/Lib/Canvas.py]: Yup, FFMPEG defaults to using h264 when libx264 is installed. That's the reason for the -pix_fmt yuv420p option (they default to a pixel format that quicktime doesn't support but most other players do, so I went with the lowest common denominator).

**Comment by chaosphere2112** [comment]
Let's hang off a sec on merging this, @doutriaux1; it turns out Cisco provides a binary version of H264 that is properly licensed for anyone to use, so I'm going to try linking against that with ffmpeg and see what happens.

**Comment by chaosphere2112** [comment]
@doutriaux1 also there's all of the build failures, but that's probably not my fault.

**Comment by aashish24** [comment]
Looks like we are running into CMake version issue. This is new but I don't know why we didn't see this before.

**Comment by chaosphere2112** [comment]
@doutriaux1 Rather than porting to OpenH264's source, I'm content just sticking with libx264. It's better supported by FFMPEG, and we already made it work here. I'll push a fresh commit to restart these... not sure why they all failed.

**Comment by durack1** [comment]
@chaosphere2112 there are a bunch of `master` tweaks to get `cmake` working and other things, a rebase might be useful to pull all these little tweaks and get your branch passing all the buildbot tests..

**Comment by doutriaux1** [comment]
@chaosphere2112 looks like Ubuntu is hang. Will try on my machine before giving it the final go. Did you try under Ubuntu?

**Comment by durack1** [comment]
@chaosphere2112 to fix `travis-ci/push` you'll need to merge in #1565, it seems [garant has a standing issue with `vcs_test_click_info`](https://github.com/UV-CDAT/uvcdat/pull/1557#issuecomment-142749695) this has repeated in a couple of standing PRs #1557, #1540

**Comment by doutriaux1** [comment]
@durack1 I believe this PR is already in master so if the bot started after it went in there is no need for @chaosphere2112 to merge it in, the bots automatically merge master in.

**Comment by doutriaux1** [comment]
ooops @durack1 I take it back... the bots WILL merge it in from master, but travis is not a bot...

**Comment by doutriaux1** [comment]
@durack1 I think the click_info is an artefact of the baselines being updated for the resize merge and that code actually not making it into garant because it is not a bot.

**Comment by chaosphere2112** [comment]
@doutriaux1 @durack1 Rebased off of master, hopefully we get a clean build this time :smile:

**Comment by durack1** [comment]
@doutriaux1 @chaosphere2112 seems like `garant` just isn't getting up to date source.. Everything else is passing, I wonder if this issue is the same problem with `vcs_test_click_info` which keeps occurring, just on the one machine (`garant`)?

**Comment by chaosphere2112** [comment]
@durack1 Dunno; I can't seem to get to the cdash page to see the results.

**Comment by durack1** [comment]
@chaosphere2112 try safari, FF is getting blocked by the lab..

**Comment by jbeezley** [comment]
What are your issues with the cdash page?  We have been reconfiguring SSL on all of our servers.  If there are specific problems, I can try to work them out with our sysadmins.

**Comment by durack1** [comment]
@jbeezley it's likely an issue specific to our side (and firewall), @doutriaux1 has an open ticket with the network folks to stop blocking our connections to https://open.cdash.org

**Comment by chaosphere2112** [comment]
@durack1 @jbeezley I'm unable to get to it from any of my web browsers (or wget).

```
--2015-09-24 10:34:33--  http://open.cdash.org/
Resolving open.cdash.org... 66.194.253.26
Connecting to open.cdash.org|66.194.253.26|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://open.cdash.org/ [following]
--2015-09-24 10:34:34--  https://open.cdash.org/
Connecting to open.cdash.org|66.194.253.26|:443... connected.
Unable to establish SSL connection.
```

Odds are solid you guys upgraded to an SSL version that's too new for our firewall (same thing happened to http://djangoproject.com a while back).

Amusingly, it just happened when I tried to connect to https://OpenSSL.org

Here's the openssl s_client errors:

```
CONNECTED(00000003)
60321:error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure:/SourceCache/OpenSSL098/OpenSSL098-52.6.4/src/ssl/s23_clnt.c:593:
```

**Comment by durack1** [comment]
@chaosphere2112 can you see https://open.cdash.org/testDetails.php?test=376408935&build=4027029 using any other browsers? I'm blocked on FF41, but Safari seems to work..

**Comment by chaosphere2112** [comment]
Nope, I'm completely blocked. I'm running an old OSX on my laptop, though, so maybe it's better with a newer version of safari?

**Comment by durack1** [comment]
@chaosphere2112 I'm on `OS X 10.10.5` and `Safari 8.0.8`..

**Comment by jbeezley** [comment]
Yup, the firewall is blocking TLS 1.2.  For whatever reason, @durack1's safari is ignoring the SSL errors, which is... umm... scary.  Anyway, I got our sysadmins to enable TLS 1.0 and 1.1 on another server yesterday, so I could probably do the same here.  But you guys should really get them to support TLS 1.2.

**Comment by jbeezley** [comment]
@chaosphere2112 You actually have another problem with wget in that you are using (I think) a pre 1.0 version of openssl, which doesn't support TLS 1.2.

**Comment by jbeezley** [comment]
It isn't picking up the new `libx264`.  I'll dig around to see if I can see what is happening.  You might check if the system `libx264` is installed on the other machines.  Perhaps it is linking with the wrong one.

**Comment by jbeezley** [comment]
When I try to build manually, I get an error building x264:
```
  x264.c: In function ‘print_version_info’:

  x264.c:335:95: error: ‘X264_CSP_Iall’ undeclared (first use in this
  function)

       printf( "x264 configuration: --bit-depth=%d --chroma-format=%s\n", X264_BIT_DEPTH, chroma_format_names[X264_CHROMA_FORMAT] );
                                                                                                 ^

  x264.c:335:95: note: each undeclared identifier is reported only once for
  each function it appears in

  make[3]: *** [x264.o] Error 1
```

Looking at the configure step output, this is happening:
```
INFO] ADDITIONAL CFLAGS
CONFIGURE_ARGS IS --disable-asm;--enable-shared
LD_ARGS IS /home/jbeezley/build/add_libx264/install/lib:/home/jbeezley/build/add_libx264/install/Externals/lib:/home/jbeezley/build/add_libx264/install/Externals/lib64:/home/jbeezley/build/add_libx264/install/Externals/lib/paraview-.:
CFLAGS : -I/home/jbeezley/build/add_libx264/install/Externals/include -I/home/jbeezley/build/add_libx264/install/Externals/lib/libffi-3.1/include
configure: 513: configure: [[: not found
configure: 513: configure: linux-gnu: not found
configure: 539: configure: [[: not found
cc: error: =: No such file or directory
cc: error: icc[ .]*: No such file or directory
cc: error: ]]: No such file or directory
cc: fatal error: no input files
compilation terminated.
configure: 545: configure: [[: not found
cc: error: =: No such file or directory
cc: error: clang[ .]*: No such file or directory
cc: error: ]]: No such file or directory
cc: fatal error: no input files
compilation terminated.
configure: 1203: [: X264_CSP_Iall: unexpected operator
Config succeeded.
```

**Comment by jbeezley** [comment]
For whatever reason, the configure script is being run with `sh` rather than `bash`.  Is there a way to force running `bash ./configure`?

**Comment by chaosphere2112** [comment]
@jbeezley [This](https://github.com/UV-CDAT/uvcdat/blob/master/CMake/cdat_modules_extra/cdat_configure_step.cmake.in#L11) would be the offender. I honestly don't know why we're using `sh` there instead of `bash`... @aashish24, any notions?

**Comment by jbeezley** [comment]
[POSIX](https://www.gnu.org/software/bash/manual/bash.html#Bash-POSIX-Mode) compliance, I suppose.  But honestly if any of this actually worked on AIX or SunOS, I would be shocked.

**Comment by chaosphere2112** [comment]
Since we only support Redhat, Ubuntu, CentOS, and OSX, I think we're OK using bash. I'll try tweaking that and seeing what happens.

**Comment by chaosphere2112** [comment]
@jbeezley Looks like that broke stuff on a few platforms that were passing before... did it happen to work for you?

**Comment by jbeezley** [comment]
It looks like it fixed building libx264.  Certain configure scripts might assume POSIX `sh`.  It would be better to change the shell just for this package if there is a straightforward way to do that.

**Comment by chaosphere2112** [comment]
I’ll add a variable to the script that defaults to sh, and we can use bash for this one.

-Sam

**Comment by chaosphere2112** [comment]
@jbeezley This looks more promising; mind checking if it works for you now?

File: CMake/cdat_modules/x264_external.cmake [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ExternalProject_Add(X264
+  LIST_SEPARATOR ^^
+  DOWNLOAD_DIR ${CDAT_PACKAGE_CACHE_DIR}
+  SOURCE_DIR ${x264_source}
+  INSTALL_DIR ${x264_install}
+  URL ${X264_URL}/${X264_GZ}
+  URL_MD5 ${X264_MD5}
+  BUILD_IN_SOURCE 1
+  PATCH_COMMAND ""
+  CONFIGURE_COMMAND ${CMAKE_COMMAND} -DINSTALL_DIR=<INSTALL_DIR> -DWORKING_DIR=<SOURCE_DIR> -DCONFIGURE_ARGS=${x264_conf_args} -DCONFIGURE_SHELL=bash -P ${cdat_CMAKE_BINARY_DIR}/cdat_configure_step.cmake
```
Comment (jbeezley) [inline CMake/cdat_modules/x264_external.cmake]: I think you want `-DBASH_CONFIGURE=ON` here rather than `CONFIGURE_SHELL`.

Comment (jbeezley) [inline CMake/cdat_modules/x264_external.cmake]: Hmmm... that didn't work either.

File: CMake/cdat_modules_extra/cdat_configure_step.cmake.in [hunk 3]
```diff
@@ -1,6 +1,14 @@
 set(ENV{PKG_CONFIG_PATH} "@cdat_EXTERNALS@/lib/pkgconfig:/usr/lib/x86_64-linux-gnu/pkgconfig:/usr/lib64/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:$ENV{PKG_CONFIG_PATH}")
 set(ENV{PKG_CONFIG} "@cdat_PKG_CONFIG_EXECUTABLE@")
 
+OPTION(BASH_CONFIGURE "Whether to use BASH to configure with" OFF)
+
+if (${BASH_CONFIGURE})
```
Comment (jbeezley) [inline CMake/cdat_modules_extra/cdat_configure_step.cmake.in]: I think you also need this to be
```
if (BASH_CONFIGURE)
```

**Comment by jbeezley** [comment]
`libx264` compiles with the changes I mentioned above.

**Comment by chaosphere2112** [comment]
Made the changes. CMake and I have a rocky relationship :hankey:.

**Comment by jbeezley** [comment]
Yeah, it's one of those "the worst except for everything else" kind of things.

**Comment by chaosphere2112** [comment]
Indeed! You don't see me trying to port our build to anything else :smile:

**Comment by durack1** [comment]
@chaosphere2112 want to refresh/rebase this PR again so those tests get re-issued? I think we're close to an all green!

**Comment by chaosphere2112** [comment]
@durack1 Doing so now; fingers crossed :8ball:

**Comment by doutriaux1** [comment]
@aashish24 @sankhesh  I get this:
```
CMake Warning (dev) at CMake/cdat_modules/ffmpeg_external.cmake:13 (set):
  Policy CMP0053 is not set: Simplify variable reference and escape sequence
  evaluation.  Run "cmake --help-policy CMP0053" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  For input:

    '--disable-yasm^^--enable-gpl^^--enable-libx264^^--extra-cxxflags=@ffmpeg_source@^^--enable-shared^^--enable-zlib'

  the old evaluation rules produce:

    '--disable-yasm^^--enable-gpl^^--enable-libx264^^--extra-cxxflags=/home/doutriaux1/build_x64/build/FFMPEG^^--enable-shared^^--enable-zlib'

  but the new evaluation rules produce:

    '--disable-yasm^^--enable-gpl^^--enable-libx264^^--extra-cxxflags=@ffmpeg_source@^^--enable-shared^^--enable-zlib'

  Using the old result for compatibility since the policy is not set.
Call Stack (most recent call first):
  CMakeLists.txt:671 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.
```
We can fix this in another branch later.

**Comment by doutriaux1** [comment]
@chaosphere2112 on ubuntu I get:
```
   Command failed: 1
   '/usr/local/bin/cmake' '-DINSTALL_DIR=/home/doutriaux1/build_x64/install/Externals' '-DWORKING_DIR=/home/doutriaux1/build_x64/build/X264' '-DCONFIGURE_ARGS=--disable-asm;--enable-shared' '-DBASH_CONFIGURE=ON' '-P' '/home/doutriaux1/build_x64/CMake/cdat_configure_step.cmake'
   From /home/doutriaux1/build_x64/X264-prefix/src/X264-stamp/X264-configure-out.log

  -------------------------------------------------

  -------------------------------------------------

   From /home/doutriaux1/build_x64/X264-prefix/src/X264-stamp/X264-configure-err.log

  -------------------------------------------------

  [INFO] ADDITIONAL CFLAGS

  CONFIGURE_ARGS IS --disable-asm;--enable-shared

  LD_ARGS IS
  /home/doutriaux1/build_x64/install/lib:/home/doutriaux1/build_x64/install/Externals/lib:/home/doutriaux1/build_x64/install/Externals/lib64:/home/doutriaux1/build_x64/install/Externals/lib/paraview-.:/lgm/uvcdat/2015-09-16/lib:/lgm/uvcdat/2015-09-16/Externals/lib:


  CFLAGS : -I/home/doutriaux1/build_x64/install/Externals/include
  -I/home/doutriaux1/build_x64/install/Externals/lib/libffi-3.1/include

  bash: configure: No such file or directory

  Config Errors detected:

  

  

  CMake Error at
  /home/doutriaux1/build_x64/CMake/cdat_configure_step.cmake:25 (message):

    Error in config

  ```

I think it needs `./configure`

Will try it now.

**Comment by chaosphere2112** [comment]
@doutriaux1 I have seen the first warning message; I think all that needs to be done is convert the `@ffmpeg_source@` to `${ffmpeg_source}`. As for the second... that's weird. I wonder if the PWD isn't getting set up properly?

**Comment by chaosphere2112** [comment]
@doutriaux1 I'm not having any trouble building libx264 or ffmpeg, but I do run into problems with pip. My ubuntu VM needs a pile of updates, so I'll run those and try again.

**Comment by chaosphere2112** [comment]
Weird. Mine tried to install pip before setuptools. Worked when I tried again. Building as we speak.

**Comment by doutriaux1** [comment]
@chaosphere2112 can you please check that pip_deps refers setuptools_pkg

**Comment by chaosphere2112** [comment]
It does, which is why I found it odd.

**Comment by doutriaux1** [comment]
@chaosphere2112 configure is not in the source path:
it is at:
```
/home/doutriaux1/build_x64/build/X264/x264-snapshot-20150921-2245-stable/
```
one extra directory.

**Comment by doutriaux1** [comment]
@chaosphere2112 on my ubuntu I have this extra directory. But not on the bots... I have no idea why. @aashish24 @sankhesh could it be a cmake version difference?

**Comment by chaosphere2112** [comment]
@doutriaux1 `libx264` and `ffmpeg` both built fine on my ubuntu VM (14.04), though I think I'm missing some prereqs that are breaking the rest of the build. Unfortunately, that VM is on the computer I don't have a property pass for, so I'll have to finish debugging it tomorrow.

**Comment by chaosphere2112** [comment]
(that's before your commit)

**Comment by doutriaux1** [comment]
@chaosphere2112 don't know what's going on here. It's not crucial for RC1 (but is for 2.4) so if the other 2 PR pass I will tag RC1 tonight. Unless you get a push that fixes the bots. I won't worry about my machine for now if all bots pass your thing.

**Comment by doutriaux1** [comment]
@chaosphere2112 let's try to fix this early tomorrow morning so we can tag rc1

**Pull request merged**